src/lib/y2partitioner/dialogs/btrfs_subvolume.rb

```diff
@@ -0,0 +1,192 @@
+require "yast"
+require "cwm"
+require "y2partitioner/dialogs/popup"
+
+Yast.import "Popup"
+
+module Y2Partitioner
+  module Dialogs
+    # Popup dialog to create a btrfs subvolume
+    class BtrfsSubvolume < Popup
+      attr_reader :filesystem
+      attr_reader :form
+
+      # @param filesystem [Y2Storage::Filesystems::BlkFilesystem] a btrfs filesystem
+      def initialize(filesystem)
+        textdomain "storage"
+
+        @filesystem = filesystem
+        @form = Form.new
+      end
+
+      def title
+        _("Add subvolume")
+      end
+
+      # Shows widgets for the subvolume attributes
+      def contents
+        HVSquash(
+          VBox(
+            Left(SubvolumePath.new(form, filesystem: filesystem)),
+            Left(SubvolumeNocow.new(form, filesystem: filesystem))
+          )
+        )
+      end
+
+      # Custom layout
+      #
+      # Similar to {Popup#layout} but without help button
+      def layout
+        VBox(
+          HSpacing(50),
+          Left(Heading(Id(:title), title)),
+          VStretch(),
+          VSpacing(1),
+          VBox(ReplacePoint(Id(:contents), Empty())),
+          VSpacing(1),
+          VStretch(),
+          ButtonBox(
+            PushButton(Id(:ok), Opt(:default), Yast::Label.OKButton),
+            PushButton(Id(:cancel), Yast::Label.CancelButton)
+          )
+        )
+      end
+
+      # Form object for the dialog
+      #
+      # Widgets use this object to storage data
+      class Form
+        # @!attribute path
+        #   subvolume path
+        attr_accessor :path
+        # @!attribute nocow
+        #   subvolume nocow attribute
+        attr_accessor :nocow
+
+        def initialize
+          @path = ""
+          @nocow = false
+        end
+      end
+
+      # Input field to set the subvolume path
+      class SubvolumePath < CWM::InputField
+        attr_reader :form
+        attr_reader :filesystem
+
+        # @param form [Dialogs::BtrfsSubvolume::Form]
+        # @param filesystem [Y2Storage::Filesystems::BlkFilesystem] a btrfs filesystem
+        def initialize(form, filesystem: nil)
+          @form = form
+          @filesystem = filesystem
+        end
+
+        def label
+          _("Path")
+        end
+
+        def store
+          form.path = value
+        end
+
+        def init
+          focus
+          self.value = form.path
+        end
+
+        # Validates the path
+        #
+        # Path cannot be empty
+        # Path must start by default subvolume path
+        # Path must be uniq
+        def validate
+          valid = true
+
+          focus
+
+          if value.empty?
+            Yast::Popup.Message(_("Empty subvolume path not allowed."))
+            valid = false
+          elsif !filesystem.nil?
+            fix_path
+            if exist_path?
+              Yast::Popup.Message(format(_("Subvolume name %s already exists."), value))
+              valid = false
+            end
+          end
+
+          valid
```

Comment:
so with my suggestions how it will look like:
```
if value.empty?
  Yast::Popup.Message(_("Empty subvolume path not allowed."))
  invalid = true
elseif !filesystem.nil?
  fix_path
  if exist_path?
    Yast::Popup.Message(format(_("Subvolume name %s already exists."), value))
     invalid = true
  end
end

if invalid
  focus
  return false
else
  true
end
```
```